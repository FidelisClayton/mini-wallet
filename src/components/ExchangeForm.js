import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { css } from 'emotion'

import LabeledInput from './LabeledInput'
import Button from './Button'

import * as colors from '../helpers/colors'
import moneyFormat from '../helpers/formaters'

const styles = css({
  '.exchange-form': {
    '&__details': {
      display: 'flex',
      justifyContent: 'space-between',
      textTransform: 'uppercase',
      marginTop: '5px',
      marginBottom: '5px'
    },

    '&__label': {
      fontWeight: 'bold',
      color: colors.lightBlue
    },

    '&__value': {
      fontWeight: 'bold',
      color: colors.darkBlue
    },

    '&__form': {
      marginTop: '20px',
      paddingTop: '20px',
      borderTop: `1px solid ${colors.lightGrey}`
    },

    '&__summary': {
      display: 'flex',
      justifyContent: 'space-between',
      marginBottom: '20px',
      paddingBottom: '20px',
      marginTop: '20px',
      paddingTop: '20px',
      borderBottom: `1px solid ${colors.lightGrey}`,
      borderTop: `1px solid ${colors.lightGrey}`
    }
  }
})

const initialState = {
  data: {
    amount: 0
  },
  errors: []
}

export default class ExchangeForm extends Component {
  state = initialState

  checkAmountLimit = (value, limit) => {
    const coinName =
      this.props.type === 'sell'
        ? this.props.wallet.name
        : this.props.realWallet.name

    if (value > limit) {
      return {
        field: 'amount',
        message: `A venda não pode ser efetuada pois seu saldo de ${coinName} é insuficiente`
      }
    }

    return
  }

  handleInputChange = (inputName) => (event) => {
    const value = event.target.value

    const limit =
      this.props.type === 'sell'
        ? this.props.wallet.balance
        : this.props.realWallet.balance

    const amount =
      this.props.type === 'sell'
        ? value
        : value * this.props.prices.sell

    const errors = [
      this.checkAmountLimit(amount, limit)
    ]
      .filter(Boolean)

    this.setState({
      data: {
        ...this.state.data,
        [inputName]: event.target.value
      },
      errors
    })
  }

  getErrorForField = (fieldName) => {
    const error = this.state.errors.find(error => error.field === fieldName) || {}

    return error.message
  }

  sell = (amount, wallet, prices) => {
    const transaction = {
      fromCoin: {
        token: wallet.token,
        name: wallet.name
      },
      toCoin: {
        token: 'BRL',
        name: 'Real'
      },
      fromAmount: amount,
      toAmount: amount * prices.sell,
      toPrice: prices.sell,
      date: new Date(),
    }

    this.props.onSell(transaction)
      .then(() => this.reset(this.props.closeModal))
  }

  buy = (amount, wallet, prices) => {
    const transaction = {
      fromCoin: {
        token: this.props.realWallet.token,
        name: this.props.realWallet.name
      },
      toCoin: {
        token: wallet.token,
        name: wallet.name
      },
      fromAmount: amount * prices.sell,
      toAmount: amount,
      toPrice: prices.buy,
      date: new Date(),
    }

    this.props.onBuy(transaction)
      .then(() => this.reset(this.props.closeModal))
  }

  reset = (callback) => {
    this.setState(initialState, callback)
  }

  renderSellDetails = () => (
    <React.Fragment>
      <div className="exchange-form__details exchange-form__sell-details">
        <span className="exchange-form__label">Valor do { this.props.wallet.token }:</span>
        <span className="exchange-form__value">R$ { moneyFormat.brl(this.props.prices.sell) }</span>
      </div>

      <div className="exchange-form__details exchange-form__sell-details">
        <span className="exchange-form__label">Quantidade disponível</span>
        <span className="exchange-form__value">{ moneyFormat[this.props.wallet.token.toLowerCase()](this.props.wallet.balance) } { this.props.wallet.token }</span>
      </div>
    </React.Fragment>
  )

  renderBuyDetails = () => (
    <React.Fragment>
      <div className="exchange-form__details exchange-form__buy-details">
        <span className="exchange-form__label">Saldo disponível</span>
        <span className="exchange-form__value">R$ { moneyFormat.brl(this.props.realWallet.balance) }</span>
      </div>

      <div className="exchange-form__details exchange-form__buy-details">
        <span className="exchange-form__label">Valor do { this.props.wallet.token }:</span>
        <span className="exchange-form__value">R$ { moneyFormat.brl(this.props.prices.sell) }</span>
      </div>
    </React.Fragment>
  )

  renderSellButton = (canTrade) => (
    <Button
      className="exchange-form__sell"
      disabled={!canTrade}
      onClick={() => this.sell(Number(this.state.data.amount), this.props.wallet, this.props.prices)}
    >
      Vender
    </Button>
  )

  renderBuyButton = (canTrade) => (
    <Button
      className="exchange-form__buy"
      disabled={!canTrade}
      onClick={() => this.buy(Number(this.state.data.amount), this.props.wallet, this.props.prices)}
    >
      Comprar
    </Button>
  )

  render () {
    const {
      data: {
        amount
      }
    } = this.state

    const {
      wallet,
      prices,
      type
    } = this.props

    const canTrade =
      this.state.errors.length === 0 && this.state.data.amount > 0

    const isSelling = type === 'sell'

    return (
      <div className={`exchange-form ${styles}`}>
        { isSelling && this.renderSellDetails() }
        { !isSelling && this.renderBuyDetails() }

        <div className="exchange-form__form">
          <LabeledInput
            label={isSelling ? 'Quantidade a ser vendida' : 'Quantidade a ser comprada'}
            inputType="number"
            onChange={this.handleInputChange('amount')}
            inputProps={{
              defaultValue: amount
            }}
            error={this.getErrorForField('amount')}
          />

          <div className="exchange-form__summary">
            <span className="exchange-form__label">
              { moneyFormat[wallet.token.toLowerCase()](amount) } { wallet.token } <small>x</small> R$ { moneyFormat.brl(prices.sell) }
            </span>

            <span className="exchange-form__value">
              Total: R$ { moneyFormat.brl(amount * prices.sell) }
            </span>
          </div>

          { isSelling && this.renderSellButton(canTrade) }
          { !isSelling && this.renderBuyButton(canTrade) }
        </div>
      </div>
    )
  }
}

ExchangeForm.propTypes = {
  type: PropTypes.string.isRequired,
  wallet: PropTypes.shape({
    id: PropTypes.string,
    balance: PropTypes.number.isRequired,
    decimalPlaces: PropTypes.number.isRequired,
    exchangeable: PropTypes.bool,
    name: PropTypes.string.isRequired,
    token: PropTypes.string.isRequired
  }).isRequired,
  realWallet: PropTypes.shape({
    id: PropTypes.string,
    balance: PropTypes.number.isRequired,
    decimalPlaces: PropTypes.number.isRequired,
    exchangeable: PropTypes.bool,
    name: PropTypes.string.isRequired,
    token: PropTypes.string.isRequired
  }),
  prices: PropTypes.shape({
    buy: PropTypes.number.isRequired,
    sell: PropTypes.number.isRequired
  }),
  onBuy: PropTypes.func,
  onSell: PropTypes.func
}

ExchangeForm.defaultProps = {
  type: 'sell'
}
