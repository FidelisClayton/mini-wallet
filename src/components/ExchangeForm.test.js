import React from 'react'
import { shallow } from 'enzyme'

import ExchangeForm from './ExchangeForm'

const wallet = {
  balance: 2,
  decimalPlaces: 8,
  exchangeable: true,
  id: 'btc',
  name: 'Bitcoin',
  token: 'BTC'
}

const realWallet = {
  balance: 100000,
  decimalPlaces: 2,
  exchangeable: false,
  id: 'brl',
  name: 'Real',
  token: 'BRL'
}

const prices = {
  buy: 28896,
  sell: 29099
}

describe('Component: <ExchangeForm />', () => {
  it('should render properly', () => {
    const wrapper = shallow(
      <ExchangeForm
        wallet={wallet}
        prices={prices}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render details for sell', () => {
    const wrapper = shallow(
      <ExchangeForm
        wallet={wallet}
        prices={prices}
      />
    )

    expect(wrapper.find('.exchange-form__sell-details')).toHaveLength(2)
    expect(wrapper.find('.exchange-form__buy-details')).toHaveLength(0)
  })

  it('should render details for buy', () => {
    const wrapper = shallow(
      <ExchangeForm
        type="buy"
        wallet={wallet}
        prices={prices}
        realWallet={realWallet}
      />
    )

    expect(wrapper.find('.exchange-form__buy-details')).toHaveLength(2)
    expect(wrapper.find('.exchange-form__sell-details')).toHaveLength(0)
  })

  it('should render sell button', () => {
    const wrapper = shallow(
      <ExchangeForm
        wallet={wallet}
        prices={prices}
      />
    )

    expect(wrapper.find('.exchange-form__sell')).toHaveLength(1)
    expect(wrapper.find('.exchange-form__buy')).toHaveLength(0)
  })

  it('should render sell button', () => {
    const wrapper = shallow(
      <ExchangeForm
        type="buy"
        wallet={wallet}
        prices={prices}
        realWallet={realWallet}
      />
    )

    expect(wrapper.find('.exchange-form__buy')).toHaveLength(1)
    expect(wrapper.find('.exchange-form__sell')).toHaveLength(0)
  })

  it('should handle sell click', () => {
    const onSell = jest.fn(() => new Promise(resolve => resolve()))

    const wrapper = shallow(
      <ExchangeForm
        wallet={wallet}
        prices={prices}
        onSell={onSell}
      />
    )

    wrapper.find('.exchange-form__sell').simulate('click')

    expect(onSell.mock.calls).toHaveLength(1)
  })

  it('should handle buy click', () => {
    const onBuy = jest.fn(() => new Promise(resolve => resolve()))

    const wrapper = shallow(
      <ExchangeForm
        type="buy"
        wallet={wallet}
        prices={prices}
        onBuy={onBuy}
        realWallet={realWallet}
      />
    )

    wrapper.find('.exchange-form__buy').simulate('click')

    expect(onBuy.mock.calls).toHaveLength(1)
  })
})
