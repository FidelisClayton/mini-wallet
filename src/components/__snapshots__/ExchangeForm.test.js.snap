// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <ExchangeForm /> should render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ExchangeForm
    prices={
        Object {
            "buy": 28896,
            "sell": 29099,
          }
    }
    type="sell"
    wallet={
        Object {
            "balance": 2,
            "decimalPlaces": 8,
            "exchangeable": true,
            "id": "btc",
            "name": "Bitcoin",
            "token": "BTC",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown>
          <div
                    className="exchange-form__details exchange-form__sell-details"
          >
                    <span
                              className="exchange-form__label"
                    >
                              Valor do 
                              BTC
                              :
                    </span>
                    <span
                              className="exchange-form__value"
                    >
                              R$ 
                              29.099,00
                    </span>
          </div>
          <div
                    className="exchange-form__details exchange-form__sell-details"
          >
                    <span
                              className="exchange-form__label"
                    >
                              Quantidade disponível
                    </span>
                    <span
                              className="exchange-form__value"
                    >
                              2,00000000
                               
                              BTC
                    </span>
          </div>
</Unknown>,
        false,
        <div
          className="exchange-form__form"
>
          <LabeledInput
                    error={undefined}
                    inputProps={
                              Object {
                                        "defaultValue": 0,
                                      }
                    }
                    inputType="number"
                    label="Quantidade a ser vendida"
                    onChange={[Function]}
          />
          <div
                    className="exchange-form__summary"
          >
                    <span
                              className="exchange-form__label"
                    >
                              0,00000000
                               
                              BTC
                               
                              <small>
                                        x
                              </small>
                               R$ 
                              29.099,00
                    </span>
                    <span
                              className="exchange-form__value"
                    >
                              Total: R$ 
                              0,00
                    </span>
          </div>
          <Button
                    className="exchange-form__sell"
                    disabled={true}
                    onClick={[Function]}
          >
                    Vender
          </Button>
</div>,
      ],
      "className": "exchange-form css-k89wnn",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <div
              className="exchange-form__details exchange-form__sell-details"
>
              <span
                            className="exchange-form__label"
              >
                            Valor do 
                            BTC
                            :
              </span>
              <span
                            className="exchange-form__value"
              >
                            R$ 
                            29.099,00
              </span>
</div>,
            <div
              className="exchange-form__details exchange-form__sell-details"
>
              <span
                            className="exchange-form__label"
              >
                            Quantidade disponível
              </span>
              <span
                            className="exchange-form__value"
              >
                            2,00000000
                             
                            BTC
              </span>
</div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="exchange-form__label"
>
                  Valor do 
                  BTC
                  :
</span>,
                <span
                  className="exchange-form__value"
>
                  R$ 
                  29.099,00
</span>,
              ],
              "className": "exchange-form__details exchange-form__sell-details",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Valor do ",
                    "BTC",
                    ":",
                  ],
                  "className": "exchange-form__label",
                },
                "ref": null,
                "rendered": Array [
                  "Valor do ",
                  "BTC",
                  ":",
                ],
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "R$ ",
                    "29.099,00",
                  ],
                  "className": "exchange-form__value",
                },
                "ref": null,
                "rendered": Array [
                  "R$ ",
                  "29.099,00",
                ],
                "type": "span",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="exchange-form__label"
>
                  Quantidade disponível
</span>,
                <span
                  className="exchange-form__value"
>
                  2,00000000
                   
                  BTC
</span>,
              ],
              "className": "exchange-form__details exchange-form__sell-details",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Quantidade disponível",
                  "className": "exchange-form__label",
                },
                "ref": null,
                "rendered": "Quantidade disponível",
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "2,00000000",
                    " ",
                    "BTC",
                  ],
                  "className": "exchange-form__value",
                },
                "ref": null,
                "rendered": Array [
                  "2,00000000",
                  " ",
                  "BTC",
                ],
                "type": "span",
              },
            ],
            "type": "div",
          },
        ],
        "type": Symbol(react.fragment),
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <LabeledInput
              error={undefined}
              inputProps={
                            Object {
                                          "defaultValue": 0,
                                        }
              }
              inputType="number"
              label="Quantidade a ser vendida"
              onChange={[Function]}
/>,
            <div
              className="exchange-form__summary"
>
              <span
                            className="exchange-form__label"
              >
                            0,00000000
                             
                            BTC
                             
                            <small>
                                          x
                            </small>
                             R$ 
                            29.099,00
              </span>
              <span
                            className="exchange-form__value"
              >
                            Total: R$ 
                            0,00
              </span>
</div>,
            <Button
              className="exchange-form__sell"
              disabled={true}
              onClick={[Function]}
>
              Vender
</Button>,
            false,
          ],
          "className": "exchange-form__form",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "error": undefined,
              "inputProps": Object {
                "defaultValue": 0,
              },
              "inputType": "number",
              "label": "Quantidade a ser vendida",
              "onChange": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="exchange-form__label"
>
                  0,00000000
                   
                  BTC
                   
                  <small>
                                    x
                  </small>
                   R$ 
                  29.099,00
</span>,
                <span
                  className="exchange-form__value"
>
                  Total: R$ 
                  0,00
</span>,
              ],
              "className": "exchange-form__summary",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "0,00000000",
                    " ",
                    "BTC",
                    " ",
                    <small>
                      x
</small>,
                    " R$ ",
                    "29.099,00",
                  ],
                  "className": "exchange-form__label",
                },
                "ref": null,
                "rendered": Array [
                  "0,00000000",
                  " ",
                  "BTC",
                  " ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "x",
                    },
                    "ref": null,
                    "rendered": "x",
                    "type": "small",
                  },
                  " R$ ",
                  "29.099,00",
                ],
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Total: R$ ",
                    "0,00",
                  ],
                  "className": "exchange-form__value",
                },
                "ref": null,
                "rendered": Array [
                  "Total: R$ ",
                  "0,00",
                ],
                "type": "span",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Vender",
              "className": "exchange-form__sell",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Vender",
            "type": [Function],
          },
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown>
            <div
                        className="exchange-form__details exchange-form__sell-details"
            >
                        <span
                                    className="exchange-form__label"
                        >
                                    Valor do 
                                    BTC
                                    :
                        </span>
                        <span
                                    className="exchange-form__value"
                        >
                                    R$ 
                                    29.099,00
                        </span>
            </div>
            <div
                        className="exchange-form__details exchange-form__sell-details"
            >
                        <span
                                    className="exchange-form__label"
                        >
                                    Quantidade disponível
                        </span>
                        <span
                                    className="exchange-form__value"
                        >
                                    2,00000000
                                     
                                    BTC
                        </span>
            </div>
</Unknown>,
          false,
          <div
            className="exchange-form__form"
>
            <LabeledInput
                        error={undefined}
                        inputProps={
                                    Object {
                                                "defaultValue": 0,
                                              }
                        }
                        inputType="number"
                        label="Quantidade a ser vendida"
                        onChange={[Function]}
            />
            <div
                        className="exchange-form__summary"
            >
                        <span
                                    className="exchange-form__label"
                        >
                                    0,00000000
                                     
                                    BTC
                                     
                                    <small>
                                                x
                                    </small>
                                     R$ 
                                    29.099,00
                        </span>
                        <span
                                    className="exchange-form__value"
                        >
                                    Total: R$ 
                                    0,00
                        </span>
            </div>
            <Button
                        className="exchange-form__sell"
                        disabled={true}
                        onClick={[Function]}
            >
                        Vender
            </Button>
</div>,
        ],
        "className": "exchange-form css-k89wnn",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <div
                className="exchange-form__details exchange-form__sell-details"
>
                <span
                                className="exchange-form__label"
                >
                                Valor do 
                                BTC
                                :
                </span>
                <span
                                className="exchange-form__value"
                >
                                R$ 
                                29.099,00
                </span>
</div>,
              <div
                className="exchange-form__details exchange-form__sell-details"
>
                <span
                                className="exchange-form__label"
                >
                                Quantidade disponível
                </span>
                <span
                                className="exchange-form__value"
                >
                                2,00000000
                                 
                                BTC
                </span>
</div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="exchange-form__label"
>
                    Valor do 
                    BTC
                    :
</span>,
                  <span
                    className="exchange-form__value"
>
                    R$ 
                    29.099,00
</span>,
                ],
                "className": "exchange-form__details exchange-form__sell-details",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Valor do ",
                      "BTC",
                      ":",
                    ],
                    "className": "exchange-form__label",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Valor do ",
                    "BTC",
                    ":",
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "R$ ",
                      "29.099,00",
                    ],
                    "className": "exchange-form__value",
                  },
                  "ref": null,
                  "rendered": Array [
                    "R$ ",
                    "29.099,00",
                  ],
                  "type": "span",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="exchange-form__label"
>
                    Quantidade disponível
</span>,
                  <span
                    className="exchange-form__value"
>
                    2,00000000
                     
                    BTC
</span>,
                ],
                "className": "exchange-form__details exchange-form__sell-details",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Quantidade disponível",
                    "className": "exchange-form__label",
                  },
                  "ref": null,
                  "rendered": "Quantidade disponível",
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "2,00000000",
                      " ",
                      "BTC",
                    ],
                    "className": "exchange-form__value",
                  },
                  "ref": null,
                  "rendered": Array [
                    "2,00000000",
                    " ",
                    "BTC",
                  ],
                  "type": "span",
                },
              ],
              "type": "div",
            },
          ],
          "type": Symbol(react.fragment),
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <LabeledInput
                error={undefined}
                inputProps={
                                Object {
                                                "defaultValue": 0,
                                              }
                }
                inputType="number"
                label="Quantidade a ser vendida"
                onChange={[Function]}
/>,
              <div
                className="exchange-form__summary"
>
                <span
                                className="exchange-form__label"
                >
                                0,00000000
                                 
                                BTC
                                 
                                <small>
                                                x
                                </small>
                                 R$ 
                                29.099,00
                </span>
                <span
                                className="exchange-form__value"
                >
                                Total: R$ 
                                0,00
                </span>
</div>,
              <Button
                className="exchange-form__sell"
                disabled={true}
                onClick={[Function]}
>
                Vender
</Button>,
              false,
            ],
            "className": "exchange-form__form",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "error": undefined,
                "inputProps": Object {
                  "defaultValue": 0,
                },
                "inputType": "number",
                "label": "Quantidade a ser vendida",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="exchange-form__label"
>
                    0,00000000
                     
                    BTC
                     
                    <small>
                                        x
                    </small>
                     R$ 
                    29.099,00
</span>,
                  <span
                    className="exchange-form__value"
>
                    Total: R$ 
                    0,00
</span>,
                ],
                "className": "exchange-form__summary",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "0,00000000",
                      " ",
                      "BTC",
                      " ",
                      <small>
                        x
</small>,
                      " R$ ",
                      "29.099,00",
                    ],
                    "className": "exchange-form__label",
                  },
                  "ref": null,
                  "rendered": Array [
                    "0,00000000",
                    " ",
                    "BTC",
                    " ",
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "x",
                      },
                      "ref": null,
                      "rendered": "x",
                      "type": "small",
                    },
                    " R$ ",
                    "29.099,00",
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Total: R$ ",
                      "0,00",
                    ],
                    "className": "exchange-form__value",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Total: R$ ",
                    "0,00",
                  ],
                  "type": "span",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Vender",
                "className": "exchange-form__sell",
                "disabled": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Vender",
              "type": [Function],
            },
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
